pages/api.js

```diff
@@ -66,6 +76,70 @@ const API = ({ api, services }) => {
     resume: rate_limiting_resume
   } = rate_limiting || {};
 
+  const handleScroll = throttle(() => {
+    const currentVisibleAnchor = getVisibleAnchor();
+    if (currentVisibleAnchor !== getWindowHash()) {
+      setMenuItem(currentVisibleAnchor);
+      window.history.replaceState(
+        undefined,
+        undefined,
+        `#${currentVisibleAnchor}`
+      );
+    }
+    // approx 8 frames
+  }, 16 * 8);
+
+  const scrollToAnchor = (anchor, smooth = false) => {
+    const anchorElem = document.getElementById(anchor);
+
+    if (!anchorElem) {
+      return;
+    }
+
+    if (smooth) {
+      anchorElem.scrollIntoView({ behavior: "smooth" });
+    } else {
+      anchorElem.scrollIntoView();
+    }
+  };
+
+  const selectAnchor = anchor => {
+    setMenuItem(anchor);
+    scrollToAnchor(anchor);
+  };
+
+  const getVisibleAnchor = () => {
+    if (!contentContainer || !contentContainer.current) {
+      return;
+    }
+
+    const sectionCollection = contentContainer.current.children;
+
+    for (let i = 0; i < sectionCollection.length; i++) {
+      const isVisible = isElementVisible(sectionCollection[i]);
+      if (isVisible) {
+        return sectionCollection[i].id;
+      }
+    }
+  };
+
+  useEffect(() => {
```

Comment:
Importing a custom hook instead of defining it in place could benefit to the readability and shareability of your implementation.

In addition, my opinion is that it make sense to define these outside the api.js file as the api.js file responsibility is to aggregate content for the api page, not implement generic features.

Reply:
Sorry but I failed to produce a cleaner code organisation on this one.